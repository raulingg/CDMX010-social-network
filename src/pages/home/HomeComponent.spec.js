/* eslint-disable jest/no-identical-title */
/* eslint-disable max-len */
/* eslint-disable jest/no-focused-tests */
/* eslint-disable prefer-promise-reject-errors */
/* eslint-disable no-unused-vars */
import {
  getByAltText, getByText, queryByTestId, queryByText, screen, waitFor,
} from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import Home from './HomeComponent.js';

const observe = jest.fn();
const disconnect = jest.fn();

// you can also pass the mock implementation
// to jest.fn as an argument
window.IntersectionObserver = jest.fn(() => ({
  observe,
  disconnect,
}));

const posts = [
  {
    message: 'Post 1',
    user: 'test@example.com',
    createdAt: '2021-02-21T20:50:22.438Z',
  },
  {
    message: 'Post 2',
    user: 'test@example.com',
    createdAt: '2021-02-21T20:50:22.438Z',
  },
  {
    message: 'Post 3',
    user: 'test@example.com',
    createdAt: '2021-02-21T20:50:22.438Z',
  }];

describe('HomeComponent', () => {
  const email = 'test@example.com';
  const message = 'new post';
  const user = { email, name: 'test' };

  const create = jest.fn().mockImplementation(
    () => Promise.resolve('newuid'),
  );

  const signOut = jest.fn().mockImplementation(
    () => Promise.resolve(),
  );

  beforeEach(() => {
    document.body.innerHTML = '<div id="root"></div>';
    create.mockClear();
    signOut.mockClear();
  });

  const createPost = (messageArg = message) => {
    userEvent.type(screen.getByTestId('postBuilderInput'), messageArg);
    userEvent.click(screen.getByTestId('postBuilderButton'));
  };

  it('should render', () => {
    const target = document.getElementById('root');
    Home({ firebaseClient: {}, user }).render(target);

    expect(target.innerHTML).toMatchSnapshot();
  });

  it('should show a list of posts', () => {
    const target = document.getElementById('root');
    const home = Home({ firebaseClient: {}, user });
    home.render(target);
    home.setData({ posts });
    home.showPosts();

    expect(target.innerHTML).toMatchSnapshot();
  });

  it('should sign out', async () => {
    const target = document.getElementById('root');
    Home({ firebaseClient: { auth: { signOut } }, user }).render(target);

    userEvent.click(screen.getByText('Logout'));
    expect(signOut).toHaveBeenCalled();
  });

  describe('createPost', () => {
    it('should show a validation error message when a user submits an empty post', () => {
      const target = document.getElementById('root');
      Home({ firebaseClient: { post: { create } }, user }).render(target);
      createPost('');
      expect(screen.getByTestId('feedbackMessage').innerHTML).toBe('Mensaje es requerido!');
    });

    it('should create a new post', async () => {
      const target = document.getElementById('root');
      Home({ firebaseClient: { post: { create } }, user }).render(target);
      createPost();

      expect(create).toHaveBeenCalledWith({ message, user: email });

      await waitFor(() => {
        expect(screen.getByTestId('feedbackMessage').innerHTML).toBe('Message sent successfully!');
        expect(document.querySelector('[data-id=newuid]')).toBeInTheDocument();
      });
    });
  });

  describe('editablePost', () => {
    it('should show an editable post focused and hide post when user clicks over a post', async () => {
      const target = document.getElementById('root');
      Home({ firebaseClient: { post: { create } }, user }).render(target);
      createPost();

      await waitFor(() => {
        userEvent.click(screen.getByText(message));
        const editableMessageElement = screen.getByText(message);
        expect(queryByText(editableMessageElement.parentElement, email)).not.toBeInTheDocument();
        expect(editableMessageElement).toHaveFocus();
        expect(editableMessageElement.parentElement).toMatchSnapshot();
      });
    });

    it('should hide editable post and reshow the post when a user cancels the edition', async () => {
      const target = document.getElementById('root');
      Home({ firebaseClient: { post: { create } }, user }).render(target);
      createPost();

      await waitFor(() => {
        userEvent.click(screen.getByText(message));
        const editableElement = screen.getByText(message);
        userEvent.click(screen.getByText('Cancel'));
        const { parentElement } = screen.getByText(message);
        expect(queryByText(parentElement, email)).toBeVisible();
        expect(parentElement).not.toContainElement(editableElement);
      });
    });

    it('should hide editable post and reshow the post when user types Escape', async () => {
      const target = document.getElementById('root');
      Home({ firebaseClient: { post: { create } }, user }).render(target);
      createPost();

      await waitFor(() => {
        userEvent.click(screen.getByText(message));
        const editableElement = screen.getByText(message);
        userEvent.type(screen.getByText(message), '{esc}');
        const { parentElement } = screen.getByText(message);
        expect(queryByText(parentElement, email)).toBeVisible();
        expect(parentElement).not.toContainElement(editableElement);
      });
    });
  });
});
