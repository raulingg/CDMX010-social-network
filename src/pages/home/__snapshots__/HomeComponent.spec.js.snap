// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeComponent editablePost should show an editable post focused and hide post when user clicks over a post 1`] = `
<form
  data-id="newuid"
>
  
    
  <textarea
    autofocus=""
    cols="50"
    data-behavior="onEnableUpdateButton"
    name="message"
    rows="4"
  >
    new post
  </textarea>
  
    
  <button
    data-behavior="onUpdatePost"
    disabled=""
  >
    Edit
  </button>
  
    
  <button
    data-behavior="onCancel"
  >
    Cancel
  </button>
  
  
</form>
`;

exports[`HomeComponent should render 1`] = `
"
    <h1>Home</h1>
    
  <ul>
      <li><a href=\\"#\\" id=\\"home\\">Home</a></li>
      <li><a href=\\"#\\" id=\\"signOut\\">Logout</a></li>
  </ul>

    
  <form id=\\"postBuilderForm\\" data-testid=\\"postBuilderForm\\">
    <textarea name=\\"postBuilderInput\\" rows=\\"4\\" cols=\\"50\\" autofocus=\\"\\" data-testid=\\"postBuilderInput\\"></textarea>
    <input type=\\"submit\\" value=\\"Post\\" data-testid=\\"postBuilderButton\\">
  </form>
  <p id=\\"feedbackMessage\\" data-testid=\\"feedbackMessage\\" hidden=\\"\\">Message sent successfully!</p>

    <div id=\\"wall\\" data-testid=\\"wall\\">
      <p>Fetching posts...</p>
    </div>
    <div id=\\"observer\\"></div>
  "
`;

exports[`HomeComponent should show a list of posts 1`] = `
"
    <h1>Home</h1>
    
  <ul>
      <li><a href=\\"#\\" id=\\"home\\">Home</a></li>
      <li><a href=\\"#\\" id=\\"signOut\\">Logout</a></li>
  </ul>

    
  <form id=\\"postBuilderForm\\" data-testid=\\"postBuilderForm\\">
    <textarea name=\\"postBuilderInput\\" rows=\\"4\\" cols=\\"50\\" autofocus=\\"\\" data-testid=\\"postBuilderInput\\"></textarea>
    <input type=\\"submit\\" value=\\"Post\\" data-testid=\\"postBuilderButton\\">
  </form>
  <p id=\\"feedbackMessage\\" data-testid=\\"feedbackMessage\\" hidden=\\"\\">Message sent successfully!</p>

    <div id=\\"wall\\" data-testid=\\"wall\\">
<article class=\\"post\\" data-id=\\"undefined\\">
  <p data-behavior=\\"onEditPost\\" style=\\"white-space: pre-wrap\\">Post 1</p>
  <p>test@example.com</p>
  <p>2021-02-21T20:50:22.438Z</p>
  <button data-behavior=\\"onRemovePost\\">X</button>
  <br>
  <br>
</article>

<article class=\\"post\\" data-id=\\"undefined\\">
  <p data-behavior=\\"onEditPost\\" style=\\"white-space: pre-wrap\\">Post 2</p>
  <p>test@example.com</p>
  <p>2021-02-21T20:50:22.438Z</p>
  <button data-behavior=\\"onRemovePost\\">X</button>
  <br>
  <br>
</article>

<article class=\\"post\\" data-id=\\"undefined\\">
  <p data-behavior=\\"onEditPost\\" style=\\"white-space: pre-wrap\\">Post 3</p>
  <p>test@example.com</p>
  <p>2021-02-21T20:50:22.438Z</p>
  <button data-behavior=\\"onRemovePost\\">X</button>
  <br>
  <br>
</article>
</div>
    <div id=\\"observer\\"></div>
  "
`;
